contracts/clearing/TradingApp.sol

```diff
@@ -90,14 +32,55 @@ contract TradingApp is IForceMoveApp {
 
 		bytes memory candidateData = candidate.variablePart.appData;
 		// settlement
-		// TODO: unsure whether we should check the proof when consensus is reached
-		if (candTurnNum % 2 == 0 && proof.length == 0) {
+		if (candTurnNum % 2 == 0 && proof.length != 0) {
 			Consensus.requireConsensus(fixedPart, proof, candidate);
-			// NOTE: used just to check the data structure validity
-			ITradingStructs.Settlement memory _unused = abi.decode(
+			// Check the settlement data structure validity
+			ITradingTypes.Settlement memory _unused = abi.decode(
 				candidateData,
-				(ITradingStructs.Settlement)
+				(ITradingTypes.Settlement)
+			);
+
+			VariablePart memory prevProof = proof[proof.lenght - 1].variablePart;
+			ITradingTypes.Order memory prevOrder = abi.decode(
```

Comment:
States in `proof` are ordered in ascendingly, meaning `proof[proof.length - 1]` has `turnNum = candidate.variablePart.turnNum`. This also means that the last state in proof should be `orderResponse`, not `order`.

Reply:
Also, considering the fact that for each Order there should be OrderResponse, we can safely assume that when `candidate` is Settlement, proof.length should be even. However, the same rule holds when `candidate` is Order. This means we need to differentiate those 2 somehow...

Reply:
Unfortunately, we CAN'T state that if `proof.length == 2`, then `candidate` is Order, as Settlement can happen after just 1 order too.

Reply:
Order and Settlement states can be differentiated by number of signatures in them. Order have only 1, whereas Settlement - 2.

Reply:
If this changes in future, we can fall back to checking the internal structure of the data, basically trying to decode Order. If successful, then it is indeed order, and Settlement otherwise.